docs/design-approaches.md

```diff
@@ -0,0 +1,235 @@
+
+### Vsock
+
+Vsock is experimental in Firecracker and final implementation may influence runtime → agent communication. This doc relies on firecracker `0.12` vsock implementation. As Vsocks are very similar to regular sockets, any communication protocol can be chosen (ttrpc, grpc, http, ...) for exchanging information.
+
+
+# Design approaches
+
+* Orchestrator
+* Runtime behavior
+* Agent behavior
+* Multiplexing
+* Implement drive management for Firecracker to mount multiple containers, add drive matching functionality to Agent.
+* Extend Agent service to work with multiple containers.
+
+![design-sequence-diagram](img/design-sequence-diagram.png)
+
+
+## Orchestrator
+
+The orchestrator knows about task definitions (e.g. aware about high level picture), makes decisions based on given information, and submits requests to containerd.
+
+The orchestrator talks to all subsystems using containerd's [APIs](https://github.com/containerd/containerd/tree/master/api). In order to manage containers lifecycle it can use [task](https://github.com/containerd/containerd/tree/master/services/tasks) service, extra configuration can be specified thought option [fields](https://github.com/containerd/containerd/blob/master/api/services/tasks/v1/tasks.proto#L74). Missing functionality (like firecracker control service) can be added as an extension to `containerd`. Containerd supports GRPC plugins, so FC control can reuse existing GRPC [server](https://github.com/containerd/containerd/blob/a15b6e2097c48b632dbdc63254bad4c62b69e709/services/server/server.go#L78) and live in same process as `containerd`. Orchestrator's client uses same GRPC connection (`containerd` has `containerd.NewWithConn` to use existing connection [object](https://github.com/containerd/containerd/blob/09da2d867aff8e0bff40dbb94c539e44529aa192/client.go#L144)) for both `containerd` and FC control service. FC control and orchestrator share the same protobuf definition for client/server communication.
+
+A few details on Orchestrator / Runtime responsibilities:
+
+* Runtime/FC control manages Firecracker microVMs lifecycle e.g. creates new VM instances for new create container requests, knows how to reuse existing VMs, and decides when to shutdown instances.
+* Orchestrator uses a `VMID` for each container group and includes it in requests to Runtime to identify which microVM instance to use for serving a request. Runtime is responsible for matching given `VMID` with corresponding VM instance.
+* Orchestrator is able to get running microVM instances and query specific VM through containerd's service layer (get metrics, set metadata, get logs, etc) by `VMID`.
+* Runtime is responsible for tracking containers states and for VM termination if there are no running containers left.
+* Orchestrator is responsible for pulling container images via `containerd` client.
+
+
+## Runtime behavior
```

Comment:
Between our two PRs, whoever merges second should probably just link this Runtime section and the Agent section to the shim-design doc for more context on the final decision that was made

Reply:
Yep. I'll update this section with ref to your doc as soon as it gets merged.

Reply:
I think you should just link to Erik's doc.